### DISCUSSION TITLE
Create the concept of a not-modifying method

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5905] [Language Ideas]
The idea is that methods (including "methods", property getters/setters, event adders/removers) can be marked as not-modifying. This means that method is prohibited from making any changes to state - instance and static fields; and from calling any methods not marked as such.
```cs
notmodify bool IsHeavierThanMe(int weight) => weight > this._weight;
notmodify bool IsLighterThanMe(int weight) { this._weight -= 1; return weight < this._weight; } // compiler error

x.IsHeavierThanMe(5); // info message: this method does not modify state, and its return value is not used, 
// so consider removing this call
```
Other than finding useless method calls (which is a minor performance optimization done by the programmer), it would give the runtime more information on what the method is going to do, potentially helping it optimize even further. Don't quote me on the latter, I don't know enough to state this is useful with certainty. Also, it would allow Visual Studio debugger to evaluate these methods without having to push the "re-calc" button - it will know which methods are OK to auto-evaluate.
Another possibility is for the compiler to automatically deduce if a method is a not-modifying one, and marks it when emitting the result, as needed.

**Comment by CyrusNajmabadi** [comment]
> Another possibility is for the compiler to automatically deduce if a method is a not-modifying one, and marks it when emitting the result, as needed.

This is something very dangerous because it means that modifying a method body can update something externally visible in your signature that you dn't have control over.  Consider if you have method X and it is implicitly non-modifying (and thus marked as such when the compiler emits it).  You have it called from non-modifying method Y in some other lib (potentially in 3rd party code).  Now, if you make X modifying in the future, that will break a downstream consumer, even though you never explicitly wrote down a contract stating your method was safe to call from non-modifying consumers.

**Reply by TahirAhmadov** [reply]
I agree, and this falls under the umbrella of a number of these dilemmas; and I usually side with "explicit" declaration approach. I mentioned the automatic approach because it would allow the entirety of BCL and other large code bases to be recompiled and voila - we get the benefit.
To stick with explicit keyword approach, perhaps VS can add a refactor (or a custom refactor can be added) which does the auto-check and shows a message: this method does not modify, mark it as such, as a good practice.

**Reply by TahirAhmadov** [reply]
Perhaps we can adopt the automatic approach in some cases? For example, the vast majority of property getters just return a field, and especially all the auto-properties, can be marked as such by default.

**Reply by TahirAhmadov** [reply]
Perhaps, the rule is, property getters are marked as not-modify by default; to be able to modify state in a property getter, it has to be marked as "modifying" - using yet another keyword. That would greatly reduce the bloat. On the other hand, all other methods - "methods", setters, and events, are "modifying" by default, and have to be marked as non-modifying explicitly, to change it.

**Reply by CyrusNajmabadi** [reply]
> Perhaps we can adopt the automatic approach in some cases? For example, the vast majority of property getters just return a field, and especially all the auto-properties, can be marked as such by default.

That fits into my point though.  They're non-modifying *today*.  But the reason they are properties is so that hte impl can choose to change to something else in the future (For example, computing and caching a result).  That would make it now 'modifying' which would break all downstream consumers who depended on the implicit non-modifying marker being present.

It is fundamentally not safe to implicitly add things that will break people downstream if you make a change in the body.

**Reply by CyrusNajmabadi** [reply]
>  to be able to modify state in a property getter, it has to be marked as "modifying" 

That would be a breaking change of gargantuan proportions :)

**Reply by TahirAhmadov** [reply]
The benefit of defaulting property getters to not-modify, is that making the modify can be added as a rule violation, which has to be reviewed and suppressed; we can all agree that it's highly unusual and probably bad practice to modify state in a property getter.

**Reply by CyrusNajmabadi** [reply]
> we can all agree that it's highly unusual and probably bad practice to modify state in a property getter.

I disagree on both counts.  It's totally normal and reasonable to modify state in a property getter.  roslyn is full of this.  Consider even the trivial pattern of checking if a value is already computed, and if not computing and caching it, then returning the cached value.  This is a super normal and reasonable pattern that would now break.

**Reply by TahirAhmadov** [reply]
> That would be a breaking change of gargantuan proportions :)

You are right, of course - my brain fart. OK, back to "opt-in" for everything.

> But the reason they are properties is so that hte impl can choose to change to something else in the future (For example, computing and caching a result). That would make it now 'modifying' which would break all downstream consumers who depended on the implicit non-modifying marker being present.

Regarding this, it is a perfectly valid point. To solve this, we can use a "local field". Basically, if the only modification is caching, then the property getter can define a field in its scope, and modifying that field does not count as a modification.

**Reply by CyrusNajmabadi** [reply]
>  is that making the modify can be added as a rule violation, which has to be reviewed and suppressed;

I feel like you can potentially accomplish all of this then with an attribute and analyzer which disallows things that go against your rules here.  It would have no runtime benefit, but i'm not sure what runtime benefits come from baking this in at the language either.

**Reply by CyrusNajmabadi** [reply]
> To solve this, we can use a "local field". Basically, if the only modification is caching, then the property getter can define a field in its scope, and modifying that field does not count as a modification.

I genuinely don't get why that doesn't count as a modification.  What *is* a modification in your mind?  And why is it important to not have certain types of modifications, but allow others?  What benefit do you get from that?

**Reply by TahirAhmadov** [reply]
> I genuinely don't get why that doesn't count as a modification. What is a modification in your mind? And why is it important to not have certain types of modifications, but allow others? What benefit do you get from that?

Because that locally scoped field cannot be accessed by anything else, therefore it doesn't alter the state of the object from a "logical" perspective.
```cs
class Person
{
  string _first, _last;
  string FullName
  {
    field string? fullName;
    readonly get
    {
      if(fullName == null) fullName = this._first + " " + this._last;
      return fullName;
    }
// The more interesting question is, how do we "invalidate" this cache (ie., set that fullName field to null) if _first is changed, for example. 
// Maybe:
    private set { fullName = null; }
  }
  string FirstName { set { this._first = value; this.FullName = null!; } }
}
```

**Reply by CyrusNajmabadi** [reply]
But the property could mutate the field on each successive property call. So you'd have a non mutating object that would observably change.

**Reply by TahirAhmadov** [reply]
> But the property could mutate the field on each successive property call. So you'd have a non mutating object that would observably change.

True. From that perspective, I guess we'll have to apply strict rules - any modification, and the getter cannot be marked as "readonly" / "pure" / "not-modifying".

**Reply by TahirAhmadov** [reply]
Or create a special syntax for the caching.

**Comment by bernd5** [comment]
Isn't it the same like we have already for `struct`? There you get a readonly `this`...

e.g.:
```csharp
struct Foo{
    int _weight;
    
    readonly bool IsHeavierThanMe(int weight) => weight > this._weight;
    readonly bool IsLighterThanMe(int weight) {
        //this._weight -= 1;
        return weight < this._weight;
    } // co   
}
```

I would like it to have for non value types, too - just to make the intent clear...
If you need to do caching - it would be possible to do with a `const_cast` like:

```csharp
struct Foo{
    int _weight;
    
    public readonly bool IsHeavierThanMe(int weight) => weight > this._weight;
    public readonly bool IsLighterThanMe(int weight)
    {
        //const_cast
        System.Runtime.CompilerServices.Unsafe.AsRef(in _weight) -= 1;
        
        return weight < this._weight;
    }
}
```

(the const_cast sample is bad - but I hope you understand what I mean...)

**Reply by TahirAhmadov** [reply]
Great point, I haven't thought of that. While the unsafe approach works, I think a more elegant solution would be a local scoped field.

**Comment by 333fred** [comment]
See https://github.com/dotnet/csharplang/discussions/776, https://github.com/dotnet/csharplang/discussions/4803, and https://github.com/dotnet/csharplang/discussions/5570.

**Reply by TahirAhmadov** [reply]
Thanks. I think #776 is #4803 are relevant, and #5570 is something different.

**Reply by 333fred** [reply]
It's similar, though it extends `pure` functions to enable a form of `constexpr`.

**Reply by TahirAhmadov** [reply]
Speaking of `constexpr`, separately from this discussion, is there any movement on implementing that?

**Reply by CyrusNajmabadi** [reply]
Nope.  No movement.

**Comment by Xyncgas** [comment]
There is the ```in``` keyword.
Also if you don't want the method to modify any states just don't do it yourself man.
I get it's hard and confusing when you are on a deadline and you are loosing yourself in the bugs and you start to blame the Object-Oriented model, but I think it's a beautiful thing, OO can be functional if you make all member methods static but the power to change objects means you are changing how the functions behaves -- each object is a lambda function, by mutating the object you are mutating the lambda's parameter, more over the ability to have side affect is an addition it's not forced up on you because you don't have to do it it's opt-in it's not a defect of the language

**Reply by Xyncgas** [reply]
it's like saying people who speaks chinese are inherently good at math but you can be good at math too by learning math in english

**Comment by Xyncgas** [comment]
There is F#

**Comment by richlander** [comment]
@Xyncgas -- Please avoid analogies that could be offensive to some folks (like discussing nationalities). Please stick closer to the topic at hand.
